librz/cons/dietline.c

```diff
@@ -24,6 +24,274 @@ typedef enum {
 	MAJOR_BREAK
 } BreakMode;
 
+/**
+ * an entry of undo. it represents either a text insertion, deletion, or both.
+ * \see undo_add_entry
+ */
+typedef struct rz_line_undo_entry_t {
+	int offset; ///< the beginning index of buffer edit.
+	char *deleted_text; ///< text to be deleted. null-terminated
+	int deleted_len; ///< the length of deleted text
+	char *inserted_text; ///< text to be inserted. null-terminated.
+	int inserted_len; ///< the length of inserted text.
+	bool continuous_next; ///< if true, redo function will continuously process the next entry.
+	bool continuous_prev; ///< if true, undo function will continuously process the previous entry.
+} RzLineUndoEntry;
+
+static inline bool is_undo_entry_valid(const RzLineUndoEntry *e) {
+	if (!e) {
+		return false;
+	}
+	if (e->offset < 0) {
+		return false;
+	}
+	if (!e->deleted_len && !e->inserted_len) {
+		return false;
+	}
+	return true;
+}
+
+static void undo_entry_free(RzLineUndoEntry *e, void *user) {
+	(void)user;
+	RZ_FREE(e->deleted_text);
+	RZ_FREE(e->inserted_text);
+}
+
+static bool undo_reset(void) {
+	if (I.enable_vi_mode || I.hud) {
+		// Undo functionality does not yet support vi_mode.
+		return true;
+	}
+	if (I.undo_vec) {
+		if (I.undo_cursor == 0 && I.undo_vec->len == 0) {
+			// no need to reset
+			return true;
+		}
```

Comment:
I would just ignore this case actually and just do a free + new anyway. I don't think we save too much by doing this and at least we are sure we don't miss any case (e.g. variables that might have a different value than expected after a reset, like `undo_continue`)

Reply:
I think you can delete this case altogether.